File: cmd/qshape/attribute.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/boringsql/qshape"
	"github.com/jackc/pgx/v5"
	"github.com/spf13/cobra"
)

type (
	attrCtx struct {
		byPosition map[int]*qshape.ParamAttribution
		// Below PG 17, ruleutils prints InitPlan outputs as `$N`, in the same
		// namespace as the query's own parameters, so those numbers are
		// ambiguous and never attributed.
		initPlanParams []int
	}

	tableRef struct {
		Schema string
		Table  string
	}

	planNode struct {
		NodeType     string          `json:"Node Type"`
		Schema       string          `json:"Schema"`
		RelationName string          `json:"Relation Name"`
		Alias        string          `json:"Alias"`
		Filter       string          `json:"Filter"`
		IndexCond    string          `json:"Index Cond"`
		HashCond     string          `json:"Hash Cond"`
		RecheckCond  string          `json:"Recheck Cond"`
		JoinFilter   string          `json:"Join Filter"`
		MergeCond    string          `json:"Merge Cond"`
		TIDCond      string          `json:"TID Cond"`
		SubplanName  string          `json:"Subplan Name"`
		Plans        json.RawMessage `json:"Plans"`
	}
)

func attributeCmd() *cobra.Command {
	var (
		inPath  string
		connStr string
		top     int
		verbose bool
	)
	cmd := &cobra.Command{
		Use:   "attribute",
		Short: "Attribute $N placeholders to table.column",
		Long: `Read a clusters.json, run EXPLAIN (GENERIC_PLAN) on each cluster's
canonical SQL, and attribute every $N placeholder to a table.column.

Every parameter gets an entry; unattributed ones are confidence:"none" rather
than aborting, and a cluster is never dropped. Writes the input to stdout
with a "params" array added to each cluster.`,
		Args: cobra.NoArgs,
		RunE: func(_ *cobra.Command, _ []string) error {
			return runAttribute(inPath, connStr, top, verbose)
		},
	}
	cmd.Flags().StringVar(&inPath, "in", "", "input clusters.json (default: stdin)")
	cmd.Flags().StringVar(&connStr, "conn", "", "PostgreSQL connection string (required)")
	cmd.Flags().IntVar(&top, "top", 0, "only attribute the top N clusters (0 = all)")
	cmd.Flags().BoolVar(&verbose, "verbose", false, "print every partially attributed cluster in the summary")
	_ = cmd.MarkFlagRequired("conn")
	return cmd
}

func runAttribute(inPath, connStr string, top int, verbose bool) error {
	var r io.Reader = os.Stdin
	if inPath != "" {
		f, err := os.Open(inPath)
		if err != nil {
			return err
		}
		defer f.Close()
		r = f
	}
	var doc clustersDoc
	if err := json.NewDecoder(r).Decode(&doc); err != nil {
		return fmt.Errorf("decode clusters.json: %w", err)
	}
	if err := validateSchemaVersion(&doc); err != nil {
		return err
	}
	doc.SchemaVersion = currentSchemaVersion

	ctx := context.Background()
	conn, err := pgx.Connect(ctx, connStr)
	if err != nil {
		return fmt.Errorf("connect: %w", err)
	}
	defer conn.Close(ctx)

	cache := newTypecastCache(conn)
	var stats attrStats
	for i := range doc.Clusters {
		if top > 0 && i >= top {
			break
		}
		c := &doc.Clusters[i]
		if c.Fingerprint == "" || c.Canonical == "" {
			continue
		}
		canonical, params, explainErr := attributeCluster(ctx, conn, cache, c.Canonical)
		// Positions refer to the re-normalised canonical, so persist it even
		// when EXPLAIN failed.
		c.Canonical = canonical
		if explainErr != nil {
			stats.explainErrors++
		}
		if len(params) == 0 {
			stats.withoutParams++
			// Drop params from an earlier run: they refer to the old canonical.
			c.Params = nil
			continue
		}
		c.Params = params
		stats.add(c.Fingerprint, params)
	}

	printAttrSummary(os.Stderr, stats, verbose)
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	return enc.Encode(doc)
}

// attributeCluster re-normalises canonical, EXPLAINs it, and returns the
// canonical it used plus one entry per parameter.
func attributeCluster(ctx context.Context, conn *pgx.Conn, cache *typecastCache, canonical string) (string, []qshape.ParamAttribution, error) {
	// Re-normalise so clusters.json written by an older qshape version picks
	// up current reshape fixes (extract-field recovery, param renumbering).
	// Fall back to the stored form if parsing fails
	if renormed, err := qshape.Normalize(canonical); err == nil {
		canonical = renormed
	}
	positions := paramPositions(canonical)
	if len(positions) == 0 {
		return canonical, nil, nil
	}
	explainSQL := castFuncParamRefs(ctx, cache, canonical)
	// PREPARE + EXPLAIN EXECUTE so Postgres sets up a parameter context
	// for the $N placeholders. Works on any PG version (GENERIC_PLAN alone
	// requires 16+, and the simple-query parser rejects bare $N otherwise).
	// Types come from the typecast pass we just ran; NULL values satisfy
	// EXECUTE's arity requirement. nparams is the highest position, which
	// equals the count only when positions are contiguous (always true after
	// renumbering; on the fallback path a gap makes PREPARE fail).
	nparams := positions[len(positions)-1]
	nulls := "NULL"
	for i := 1; i < nparams; i++ {
		nulls += ", NULL"
	}
	// force_generic_plan keeps $N in the plan output instead of inlining
	// the NULL arguments. Without it Postgres produces a custom plan with
	// `WHERE col = NULL` filters that walkPlan can't attribute
	script := "SET LOCAL plan_cache_mode = force_generic_plan;\n"
	script += "PREPARE _qshape_tmp AS " + explainSQL + ";\n"
	script += "EXPLAIN (FORMAT JSON) EXECUTE _qshape_tmp(" + nulls + ");\n"
	script += "DEALLOCATE _qshape_tmp;"
	// SET LOCAL only applies inside a transaction — wrap the whole script
	script = "BEGIN;\n" + script + "\nCOMMIT;"
	planJSON, err := readPlanJSON(ctx, conn, script)
	if err != nil {
		// A mid-batch error aborts the BEGIN'd transaction and skips the
		// trailing COMMIT in the same simple-query batch, so the connection
		// stays in aborted state and every following cluster fails with
		// 25P02. ROLLBACK resets it before the next call.
		_, _ = conn.Exec(ctx, "ROLLBACK")
		_, _ = conn.Exec(ctx, "DEALLOCATE IF EXISTS _qshape_tmp")
		return canonical, noneEntries(positions, err.Error()), err
	}
	return canonical, attributeFromPlan(planJSON, positions), nil
}

// attributeFromPlan turns an EXPLAIN (FORMAT JSON) result into one entry per
// canonical position.
func attributeFromPlan(planJSON []byte, positions []int) []qshape.ParamAttribution {
	var plans []struct {
		Plan json.RawMessage `json:"Plan"`
	}
	if err := json.Unmarshal(planJSON, &plans); err != nil {
		return noneEntries(positions, err.Error())
	}
	if len(plans) == 0 {
		return noneEntries(positions, "")
	}
	ctx := &attrCtx{byPosition: map[int]*qshape.ParamAttribution{}}
	walkPlan(plans[0].Plan, ctx)
	for _, p := range ctx.initPlanParams {
		delete(ctx.byPosition, p)
	}
	return fillPositions(ctx.byPosition, positions)
}

// readPlanJSON runs a multi-statement script (PREPARE; EXPLAIN; DEALLOCATE)
// through simple-query protocol and returns the first text cell of the
// first result-bearing statement. pgx's Query path sanitises $N against
// bound args; here we send the script verbatim
func readPlanJSON(ctx context.Context, conn *pgx.Conn, script string) ([]byte, error) {
	mrr := conn.PgConn().Exec(ctx, script)
	defer mrr.Close()
	var out []byte
	for mrr.NextResult() {
		rr := mrr.ResultReader()
		for rr.NextRow() {
			vals := rr.Values()
			if len(vals) > 0 && out == nil {
				out = append([]byte(nil), vals[0]...)
			}
		}
		if _, err := rr.Close(); err != nil {
			return nil, err
		}
	}
	if err := mrr.Close(); err != nil {
		return nil, err
	}
	if out == nil {
		return nil, fmt.Errorf("no rows returned")
	}
	return out, nil
}

// maxParamNumber returns the highest $N in sql, or 0 if none.
func maxParamNumber(sql string) int {
	hi := 0
	for _, n := range scanParams(sql) {
		hi = max(hi, n)
	}
	return hi
}

func walkPlan(raw json.RawMessage, ctx *attrCtx) {
	if len(raw) == 0 {
		return
	}
	var n planNode
	if err := json.Unmarshal(raw, &n); err != nil {
		return
	}

	// Track alias → table mapping so we can resolve `u.id = $1` to users.id.
	// Function Scan on a system view like pg_catalog.pg_settings leaves
	// RelationName empty but still sets Alias; use the alias as the table name
	// so conds like `(name = $1)` attribute to pg_settings.name. A CTE or
	// Subquery Scan alias is not a table statistics exist for, so those stay
	// heuristic.
	if _, returns, ok := strings.Cut(n.SubplanName, "(returns "); ok {
		ctx.initPlanParams = append(ctx.initPlanParams, scanParams(returns)...)
	}
	aliasToTable := map[string]tableRef{}
	var fallbackSchema, fallbackTable, fallbackConfidence string
	if n.RelationName != "" {
		t := tableRef{Schema: n.Schema, Table: n.RelationName}
		aliasToTable[n.RelationName] = t
		if n.Alias != "" && n.Alias != n.RelationName {
			aliasToTable[n.Alias] = t
		}
		fallbackSchema, fallbackTable, fallbackConfidence = n.Schema, n.RelationName, "exact"
	} else if n.Alias != "" {
		t := tableRef{Schema: n.Schema, Table: n.Alias}
		aliasToTable[n.Alias] = t
		fallbackSchema, fallbackTable, fallbackConfidence = n.Schema, n.Alias, "heuristic"
		if n.NodeType == "Function Scan" {
			fallbackConfidence = "exact"
		}
	}

	for _, cond := range []string{
		n.IndexCond, n.HashCond, n.Filter, n.RecheckCond, n.JoinFilter,
		n.MergeCond, n.TIDCond,
	} {
		if cond == "" {
			continue
		}
		attributeCond(cond, aliasToTable, fallbackSchema, fallbackTable, fallbackConfidence, ctx)
	}

	if len(n.Plans) > 0 {
		var children []json.RawMessage
		if err := json.Unmarshal(n.Plans, &children); err == nil {
			for _, c := range children {
				walkPlan(c, ctx)
			}
		}
	}
}

func attributeCond(cond string, aliases map[string]tableRef, fallbackSchema, fallbackTable, fallbackConfidence string, ctx *attrCtx) {
	forEachComparison(cond, func(left, right string) {
		l := classifyOperand(left)
		r := classifyOperand(right)
		switch {
		case l.kind == operandColumn && r.kind != operandColumn && r.kind != operandOther:
			emitComparison(l, r, aliases, fallbackSchema, fallbackTable, fallbackConfidence, ctx)
		case r.kind == operandColumn && l.kind != operandColumn && l.kind != operandOther:
			emitComparison(r, l, aliases, fallbackSchema, fallbackTable, fallbackConfidence, ctx)
		}
	})
}

// emitComparison records the parameter(s) on the non-column side against the
// column side.
func emitComparison(col, other operand, aliases map[string]tableRef, fallbackSchema, fallbackTable, fallbackConfidence string, ctx *attrCtx) {
	ref, confidence := resolveColumn(col.alias, aliases, fallbackSchema, fallbackTable, fallbackConfidence)
	column := col.column
	if ref.Table == "" {
		column = ""
	}

	a := qshape.ParamAttribution{Schema: ref.Schema, Table: ref.Table, Column: column, Confidence: confidence}
	switch other.kind {
	case operandArrayParam:
		a.Shape = "array"
	case operandExpr:
		a.Note = other.text
		if confidence == "exact" && !exprHasOtherColumn(other.text) {
			a.Confidence = "expression"
		} else {
			a.Confidence = "none"
		}
	}
	for _, pos := range other.poss {
		a.Position = pos
		ctx.record(a)
	}
}

// resolveColumn maps an operand qualifier to a table. An empty qualifier on a
// scan node is PG pinning the column to that scan (exact); a qualifier that
// doesn't resolve is a best-effort guess (heuristic).
func resolveColumn(alias string, aliases map[string]tableRef, fallbackSchema, fallbackTable, fallbackConfidence string) (tableRef, string) {
	if ref, ok := aliases[alias]; ok {
		return ref, "exact"
	}
	if fallbackTable == "" {
		return tableRef{}, "none"
	}
	ref := tableRef{Schema: fallbackSchema, Table: fallbackTable}
	if alias == "" {
		return ref, fallbackConfidence
	}
	return ref, "heuristic"
}

// confidenceRank orders attribution confidence for merging hits across plan
// nodes; "none" ranks 0.
var confidenceRank = map[string]int{"exact": 3, "expression": 2, "heuristic": 1}

func (ctx *attrCtx) record(a qshape.ParamAttribution) {
	if prev, ok := ctx.byPosition[a.Position]; ok && confidenceRank[prev.Confidence] >= confidenceRank[a.Confidence] {
		return
	}
	if a.Confidence == "none" && a.Note == "" {
		// fillPositions emits the bare none entry; don't shadow a real one.
		return
	}
	ctx.byPosition[a.Position] = &a
}

File: crates/qshape-cli/src/attribute.rs
//! Attribute $N placeholders to table.column by running EXPLAIN(GENERIC_PLAN)
//! on each cluster's canonical and walking the plan tree

use std::collections::HashMap;
use std::io::{self, Write};
use std::sync::OnceLock;

use anyhow::{Context, Result, anyhow};
use postgres::{Client, NoTls, SimpleQueryMessage};
use qshape_core::{
    CURRENT_SCHEMA_VERSION, ParamAttribution, cast_func_param_refs, max_param_number, normalize,
};
use regex::Regex;
use serde::Deserialize;

use crate::loader::load_clusters;
use crate::typecast::TypecastCache;

pub fn run(in_path: Option<&str>, conn_str: &str, top: usize) -> Result<()> {
    let mut doc = load_clusters(in_path)?;

    let mut client = Client::connect(conn_str, NoTls).context("connect")?;
    let mut cache = TypecastCache::new(&mut client);
    let mut attributed = 0usize;
    let mut skipped = 0usize;

    let limit = if top == 0 { doc.clusters.len() } else { top.min(doc.clusters.len()) };
    for c in doc.clusters.iter_mut().take(limit) {
        if c.fingerprint.is_empty() || c.canonical.is_empty() {
            continue;
        }
        match attribute_cluster(&mut cache, &c.canonical) {
            Ok(params) if params.is_empty() => skipped += 1,
            Ok(params) => {
                c.params = params;
                attributed += 1;
            }
            Err(e) => {
                skipped += 1;
                c.params = vec![ParamAttribution {
                    confidence: "none".to_string(),
                    note: format!("{e:#}"),
                    ..ParamAttribution::default()
                }];
            }
        }
    }

    eprintln!("attributed {attributed} clusters, {skipped} skipped");
    doc.schema_version = CURRENT_SCHEMA_VERSION.to_string();
    let stdout = io::stdout();
    let mut h = stdout.lock();
    serde_json::to_writer_pretty(&mut h, &doc)?;
    h.write_all(b"\n")?;
    Ok(())
}

fn attribute_cluster(
    cache: &mut TypecastCache<'_>,
    canonical: &str,
) -> Result<Vec<ParamAttribution>> {
    // re-normalize so older clusters.json picks up current reshape fixes
    let canonical = normalize(canonical).unwrap_or_else(|_| canonical.to_string());
    let explain_sql = cast_func_param_refs(&canonical, &mut cache.lookup());
    let nparams = max_param_number(&explain_sql);

    // PREPARE + EXPLAIN EXECUTE gives PG param context for $N.
    // force_generic_plan keeps $N in plan output instead of inlining
    // NULL args. SET LOCAL needs transaction, hence BEGIN/COMMIT
    let mut script = String::from(
        "BEGIN;\nSET LOCAL plan_cache_mode = force_generic_plan;\nPREPARE _qshape_tmp AS ",
    );
    script.push_str(&explain_sql);
    script.push_str(";\n");
    if nparams > 0 {
        script.push_str("EXPLAIN (FORMAT JSON) EXECUTE _qshape_tmp(");
        for i in 0..nparams {
            if i > 0 {
                script.push_str(", ");
            }
            script.push_str("NULL");
        }
        script.push_str(");\n");
    } else {
        script.push_str("EXPLAIN (FORMAT JSON) EXECUTE _qshape_tmp;\n");
    }
    script.push_str("DEALLOCATE _qshape_tmp;\nCOMMIT;");

    let plan_json = match read_plan_json(cache.client_mut(), &script) {
        Ok(v) => v,
        Err(e) => {
            // mid-batch error aborts BEGIN, reset connection
            let _ = cache.client_mut().simple_query("ROLLBACK");
            let _ = cache.client_mut().simple_query("DEALLOCATE IF EXISTS _qshape_tmp");
            return Err(e);
        }
    };

    #[derive(Deserialize)]
    struct PlanWrapper {
        #[serde(rename = "Plan")]
        plan: serde_json::Value,
    }
    let plans: Vec<PlanWrapper> =
        serde_json::from_str(&plan_json).context("parse EXPLAIN JSON")?;
    let Some(first) = plans.into_iter().next() else { return Ok(Vec::new()) };

    let mut ctx = AttrCtx { by_position: HashMap::new() };
    walk_plan(&first.plan, "", "", &mut ctx);

    let mut out: Vec<ParamAttribution> = ctx.by_position.into_values().collect();
    out.sort_by_key(|p| p.position);
    Ok(out)
}

fn read_plan_json(client: &mut Client, script: &str) -> Result<String> {
    let messages = client.simple_query(script).context("EXPLAIN script")?;
    for msg in messages {
        if let SimpleQueryMessage::Row(row) = msg
            && let Some(text) = row.get(0)
        {
            return Ok(text.to_string());
        }
    }
    Err(anyhow!("EXPLAIN returned no rows"))
}

struct AttrCtx {
    by_position: HashMap<i32, ParamAttribution>,
}

#[derive(Default)]
struct TableRef {
    schema: String,
    table: String,
}

#[derive(Deserialize, Default)]
struct PlanNode {
    #[serde(rename = "Schema", default)]
    schema: String,
    #[serde(rename = "Relation Name", default)]
    relation_name: String,
    #[serde(rename = "Alias", default)]
    alias: String,
    #[serde(rename = "Filter", default)]
    filter: String,
    #[serde(rename = "Index Cond", default)]
    index_cond: String,
    #[serde(rename = "Hash Cond", default)]
    hash_cond: String,
    #[serde(rename = "Recheck Cond", default)]
    recheck_cond: String,
    #[serde(rename = "Join Filter", default)]
    join_filter: String,
    #[serde(rename = "Merge Cond", default)]
    merge_cond: String,
    #[serde(rename = "Plans", default)]
    plans: Vec<serde_json::Value>,
}

fn walk_plan(raw: &serde_json::Value, _parent_schema: &str, _parent_table: &str, ctx: &mut AttrCtx) {
    let Ok(n) = serde_json::from_value::<PlanNode>(raw.clone()) else { return };

    // alias to table so u.id = $1 resolves to users.id. Function Scan
    // on system view leaves Relation Name empty but sets Alias — use
    // alias as table so `(name = $1)` attributes correctly
    let mut aliases: HashMap<String, TableRef> = HashMap::new();
    let mut fallback_table = n.relation_name.clone();
    let fallback_schema = n.schema.clone();
    if !n.relation_name.is_empty() {
        aliases.insert(
            n.relation_name.clone(),
            TableRef { schema: n.schema.clone(), table: n.relation_name.clone() },
        );
        if !n.alias.is_empty() && n.alias != n.relation_name {
            aliases.insert(
                n.alias.clone(),
                TableRef { schema: n.schema.clone(), table: n.relation_name.clone() },
            );
        }
    } else if !n.alias.is_empty() {
        aliases.insert(
            n.alias.clone(),
            TableRef { schema: n.schema.clone(), table: n.alias.clone() },
        );
        fallback_table = n.alias.clone();
    }

    for cond in [
        &n.index_cond,
        &n.hash_cond,
        &n.filter,
        &n.recheck_cond,
        &n.join_filter,
        &n.merge_cond,
    ] {
        if cond.is_empty() {
            continue;
        }
        attribute_cond(cond, &aliases, &fallback_schema, &fallback_table, ctx);
    }

    for child in &n.plans {
        walk_plan(child, &n.schema, &n.relation_name, ctx);
    }
}

// column op $N or $N op column, alias.column optional
const PARAM_COND_PATTERN: &str = r"(?:\(?(\w+)\.)?(\w+)\s*(?:=|<|>|<=|>=|<>|!=)\s*\$(\d+)|\$(\d+)\s*(?:=|<|>|<=|>=|<>|!=)\s*(?:\(?(\w+)\.)?(\w+)";
// column IN ($N, ...), capture only first param
const PARAM_IN_PATTERN: &str = r"(?:\(?(\w+)\.)?(\w+)\s+(?:=\s*ANY\s*\()?IN\s*\(\s*\$(\d+)";

fn param_cond_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(PARAM_COND_PATTERN).expect("static regex"))
}

fn param_in_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(PARAM_IN_PATTERN).expect("static regex"))
}

fn attribute_cond(
    cond: &str,
    aliases: &HashMap<String, TableRef>,
    fallback_schema: &str,
    fallback_table: &str,
    ctx: &mut AttrCtx,
) {
    for caps in param_cond_re().captures_iter(cond) {
        // first alt: [1]=alias [2]=col [3]=pos
        // second:    [4]=pos   [5]=alias [6]=col
        let (alias_or_table, col, pos) = match caps.get(3) {
            Some(p) => (
                caps.get(1).map_or("", |m| m.as_str()),
                caps.get(2).map_or("", |m| m.as_str()),
                p.as_str(),
            ),
            None => match caps.get(4) {
                Some(p) => (
                    caps.get(5).map_or("", |m| m.as_str()),
                    caps.get(6).map_or("", |m| m.as_str()),
                    p.as_str(),
                ),
                None => continue,
            },
        };
        record_param(alias_or_table, col, pos, aliases, fallback_schema, fallback_table, ctx);
    }
    for caps in param_in_re().captures_iter(cond) {
        let alias_or_table = caps.get(1).map_or("", |m| m.as_str());
        let col = caps.get(2).map_or("", |m| m.as_str());
        let pos = caps.get(3).map_or("", |m| m.as_str());
        record_param(alias_or_table, col, pos, aliases, fallback_schema, fallback_table, ctx);
    }
}

fn record_param(
    alias_or_table: &str,
    col: &str,
    pos_str: &str,
    aliases: &HashMap<String, TableRef>,
    fallback_schema: &str,
    fallback_table: &str,
    ctx: &mut AttrCtx,
) {
    let Ok(pos) = pos_str.parse::<i32>() else { return };
    if let Some(existing) = ctx.by_position.get(&pos)
        && existing.confidence == "exact"
    {
        return;
    }

    let (schema, table, confidence) = if let Some(t) = aliases.get(alias_or_table) {
        (t.schema.clone(), t.table.clone(), "exact")
    } else if !fallback_table.is_empty() {
        // unqualified column on scan node = exact (PG telling us
        // which scan column belongs to). Downgrade only when
        // qualifier failed to resolve
        let conf = if alias_or_table.is_empty() { "exact" } else { "heuristic" };
        (fallback_schema.to_string(), fallback_table.to_string(), conf)
    } else {
        (String::new(), String::new(), "none")
    };

    ctx.by_position.insert(
        pos,
        ParamAttribution {
            position: pos,
            schema,
            table,
            column: col.to_string(),
            confidence: confidence.to_string(),
            note: String::new(),
        },
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ctx() -> AttrCtx {
        AttrCtx { by_position: HashMap::new() }
    }

    fn aliases<const N: usize>(items: [(&str, &str, &str); N]) -> HashMap<String, TableRef> {
        items
            .into_iter()
            .map(|(k, schema, table)| {
                (k.to_string(), TableRef { schema: schema.to_string(), table: table.to_string() })
            })
            .collect()
    }

    #[test]
    fn aliased_equal() {
        let mut c = ctx();
        let a = aliases([("u", "auth", "user_account")]);
        attribute_cond("(u.user_id = $1)", &a, "auth", "user_account", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed");
        assert_eq!(p.schema, "auth");
        assert_eq!(p.table, "user_account");
        assert_eq!(p.column, "user_id");
        assert_eq!(p.confidence, "exact");
    }

    // PG emits bare column names in plan text when scan is unambiguous.
    // Plan node pins column to its relation — exact, not a guess
    #[test]
    fn unqualified_on_scan_is_exact() {
        let mut c = ctx();
        let a = aliases([("session", "auth", "session")]);
        attribute_cond("(id = $1)", &a, "auth", "session", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed");
        assert_eq!(p.table, "session");
        assert_eq!(p.column, "id");
        assert_eq!(p.confidence, "exact");
    }

    // Qualifier that doesn't resolve (outer-scope ref, schema-qualified
    // name, subplan name) — best-guess to current relation as heuristic
    #[test]
    fn mismatched_qualifier_is_heuristic() {
        let mut c = ctx();
        attribute_cond("(outer_alias.id = $1)", &HashMap::new(), "auth", "session", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed");
        assert_eq!(p.confidence, "heuristic");
    }

    #[test]
    fn multiple_params() {
        let mut c = ctx();
        let a = aliases([("t", "auth", "oauth_token")]);
        attribute_cond(
            "((t.access_sha = $2) AND (t.access_hash = hashtext($1)))",
            &a,
            "auth",
            "oauth_token",
            &mut c,
        );

        let p2 = c.by_position.get(&2).expect("param 2 attributed");
        assert_eq!(p2.column, "access_sha");
        // $1 wrapped in hashtext(...) — no direct column comparison, OK
        // either way that the regex misses it
    }

    // PG plans system views like pg_catalog.pg_settings as Function Scan.
    // Plan node has no "Relation Name" but carries view name in Alias —
    // walk_plan must still attribute conds on that node
    #[test]
    fn function_scan_with_alias_only() {
        let plan: serde_json::Value = serde_json::from_str(
            r#"{
                "Node Type": "Function Scan",
                "Function Name": "pg_show_all_settings",
                "Alias": "pg_settings",
                "Filter": "(name = $1)"
            }"#,
        )
        .unwrap();
        let mut c = ctx();
        walk_plan(&plan, "", "", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed via Alias fallback");
        assert_eq!(p.table, "pg_settings");
        assert_eq!(p.column, "name");
        assert_eq!(p.confidence, "exact");
    }

    #[test]
    fn preserves_exact_over_heuristic() {
        let mut c = ctx();
        let a = aliases([("u", "auth", "user_account")]);
        attribute_cond("(u.user_id = $1)", &a, "", "", &mut c);
        // second hit that would be heuristic on different relation
        attribute_cond("(user_id = $1)", &HashMap::new(), "public", "other_table", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed");
        assert_eq!(p.table, "user_account");
        assert_eq!(p.confidence, "exact");
    }

    // IN-list pattern: column IN ($N, $M, ...) attributes only first param
    #[test]
    fn in_list_attributes_first_param() {
        let mut c = ctx();
        let a = aliases([("u", "auth", "user_account")]);
        attribute_cond("(u.id IN ($3, $4, $5))", &a, "auth", "user_account", &mut c);

        let p = c.by_position.get(&3).expect("param 3 attributed");
        assert_eq!(p.column, "id");
        assert_eq!(p.confidence, "exact");
        assert!(c.by_position.get(&4).is_none());
    }

    // Reverse form: $N op column instead of column op $N
    #[test]
    fn reverse_form_param_first() {
        let mut c = ctx();
        let a = aliases([("u", "auth", "user_account")]);
        attribute_cond("($1 = u.email)", &a, "auth", "user_account", &mut c);

        let p = c.by_position.get(&1).expect("param 1 attributed");
        assert_eq!(p.column, "email");
        assert_eq!(p.confidence, "exact");
    }

    // walk_plan recurses into nested Plans
    #[test]
    fn walk_plan_descends_into_children() {
        let plan: serde_json::Value = serde_json::from_str(
            r#"{
                "Node Type": "Hash Join",
                "Hash Cond": "(orders.user_id = users.id)",
                "Plans": [
                    {
                        "Node Type": "Seq Scan",
                        "Relation Name": "users",
                        "Alias": "users",
                        "Filter": "(email = $1)"
                    }
                ]
            }"#,
        )
        .unwrap();
        let mut c = ctx();
        walk_plan(&plan, "", "", &mut c);

        let p = c.by_position.get(&1).expect("param 1 from child Seq Scan");
        assert_eq!(p.table, "users");
        assert_eq!(p.column, "email");
    }
}
